### ISSUE TITLE
importprivkey with rescan causes nodes which attempt to connect during rescan to be banned

### ISSUE THREAD

**Issue by Justaphf** [gandrewstone/bitcoin#718]
Rescan is a blocking, long-running operation that freezes the client for > 45 minutes (on an i5 PC, SSD).

During this "freeze":
* No new transactions are processed
* No new blocks are processed
* No new client connections are processed

Once the rescan ends, the backlog of transactions, blocks, and client connection requests are quickly processed, but due to the "too many connection attempts in too short a time" ban logic, any nodes that tried to connect to the node more than 4 times during the entire rescan time period will be banned due to all of the pending connection requests being processed within a second of each other, regardless of when the connection request was initiated.

Example log snippet:
```
2017-08-01 00:44:03 Still rescanning. At block 183416. Progress=0.009552
...
2017-08-01 01:32:03 Still rescanning. At block 477796. Progress=0.998709
2017-08-01 01:32:16 Banning xxx.xxx.xxx.xxx:yyyyy for 4 hours: Too many connection attempts - connection dropped
2017-08-01 01:32:16 Banning xxx.xxx.xxx.xxx:yyyyy for 4 hours: Too many connection attempts - connection dropped
2017-08-01 01:32:16 Banning xxx.xxx.xxx.xxx:yyyyy for 4 hours: Too many connection attempts - connection dropped
2017-08-01 01:32:16 Banning xxx.xxx.xxx.xxx:yyyyy for 4 hours: Too many connection attempts - connection dropped
2017-08-01 01:32:16 Banning xxx.xxx.xxx.xxx:yyyyy for 4 hours: Too many connection attempts - connection dropped
2017-08-01 01:32:16 Banning xxx.xxx.xxx.xxx:yyyyy for 4 hours: Too many connection attempts - connection dropped
2017-08-01 01:32:16 Banning xxx.xxx.xxx.xxx:yyyyy for 4 hours: Too many connection attempts - connection dropped
2017-08-01 01:32:16 Banning xxx.xxx.xxx.xxx:yyyyy for 4 hours: Too many connection attempts - connection dropped
2017-08-01 01:32:29 Acceptable block: ver:20000002 time:1501551058 size: 269987 Tx:343 Sig:882
2017-08-01 01:32:29 Acceptable block: ver:20000002 time:1501548573 size: 926151 Tx:1026 Sig:1880
2017-08-01 01:32:30 Acceptable block: ver:20000002 time:1501549123 size: 22856 Tx:54 Sig:84
2017-08-01 01:32:30 Acceptable block: ver:20000002 time:1501550164 size: 537530 Tx:422 Sig:649
2017-08-01 01:32:30 Acceptable block: ver:20000002 time:1501548573 size: 926151 Tx:1026 Sig:1880
2017-08-01 01:32:30 Acceptable block: ver:20000002 time:1501548573 size: 926151 Tx:1026 Sig:1880
2017-08-01 01:32:30 Acceptable block: ver:20000002 time:1501550027 size: 664970 Tx:1545 Sig:2937
2017-08-01 01:32:30 Acceptable block: ver:20000002 time:1501548573 size: 926151 Tx:1026 Sig:1880
```

**Comment by sickpig** [comment]
@Justaphf thanks for the report. 
Which version of BU are you using? 1.0.3 or BUcash 1.1.0.0?

**Comment by Justaphf** [comment]
I initially noticed this on the dev branch.  The log I included above came from the BitcoinCash branch.  I have not tested release.

I plan to do a little further investigation to see if i can reproduce in a controlled environment, it's possible that the bans were actually legitimate, but they just didn't get logged until rescan completed.

**Comment by sickpig** [comment]
@Justaphf thanks for your willingness to further look into the issue. Please keep us posted.

**Comment by Justaphf** [comment]
@sickpig I've confirmed, via the test method listed below, that while performing a rescan after key import multiple peer connection attempts end up tripping the too many connection attempts ban.

Test procedure:
1. On node A import a private key and begin a rescan.
2. On node B, using the watch command cause a oneshot addnode to be attempted every 7 minutes
`watch -d -n 420 'bitcoin-cli addnode "nodeA:8333" oneshot'`
3. Upon completion of the rescan on node A, notice that node B has been banned on node A as soon as the rescan ends.

For reference, I also ran the watch command attempting to oneshot connect every 7 minutes while a rescan was not in progress, both with and without a secondary watch disconnecting node B from node A every 7 minutes offset by 1 minute from the addnode attempt and left it running for an hour.  In either case, no ban was created, so this didn't trip the too many connection attempts.

Most likely while the rescan is ongoing all incoming node connection attempts are blocking on one of the locks and then they all get released in quick succession once the rescan ends and frees the needed lock.  I'm not sure what lock it is, but based on the `AcceptConnection` code it is probably `cs_vNodes`, as this lock blocks just before the too many connection attempts check.

**Comment by sickpig** [comment]
@Justaphf 
thank you very much for the detailed report and for writing down how to reproduce the problem. Indeed it seems we have something to fix here.

**Comment by sickpig** [comment]
@Justaphf this seems related to this Core issue: 

https://github.com/bitcoin/bitcoin/issues/10987

Quoting @sipa:

`When the wallet is rescanned at startup in bitcoind, there seems to be no way to interrupt that in a clean way (other than forcefully killing the process).`

**Comment by Justaphf** [comment]
@sickpig Yes, the long blocking time when rescanning is a known issue, however the banning is, AFAIK, only an issue in BU since I don't think any other clients are using the too many connections in too short a time ban logic.

The ideal fix would be to correct the blocking during rescan, though I imagine that will be a much harder task as it likely traces back to the overuse of `cs_main` locks all over the place.

Two potential short-term fixes for just the banning issue may be to either fail out of AcceptConnection immediately while a rescan is ongoing (assuming it is blocking inside this function and not somewhere earlier), or to temporarily disable the too many connections ban logic while rescanning and re-enabling after the backlog of connection attempts has been cleared on completion of the rescan.

**Comment by Justaphf** [comment]
Was fixed in #799.  Closing.
